### DISCUSSION TITLE
How do Options impact Validate

### DISCUSSION THREAD

**Discussion by verdverm** [cue-lang/cue#1329] [Q&A]
I have the following, I expected no errors for `loose` and errors for `every`. How are the Options used by Validate? It seems several make no difference. (Definitions, Hidden, Optional)

```go
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/errors"
)

const val = `
i: int
s: string
t: [string]: string
_h: int
_h: "hidden"
#d: int
#d: "bar"
`

func main() {
	c := cuecontext.New()
	v := c.CompileString(val)

	//   try out different validation schemes
	printErr("loose error", loose(v))
	printErr("every error", every(v))
	printErr("strict error", strict(v))

	fmt.Printf("\nvalue:\n%#v\n", v)
}

func printErr(prefix string, err error) {
	if err != nil {
		msg := errors.Details(err, nil)
		fmt.Printf("%s:\n%s\n", prefix, msg)
	}
}

func loose(v cue.Value) error {
	return v.Validate(
		// not final or concrete
		cue.Concrete(false),
		// check minimally
		cue.Definitions(false),
		cue.Hidden(false),
		cue.Optional(false),
	)
}

func every(v cue.Value) error {
	return v.Validate(
		// not final or concrete
		cue.Concrete(false),
		// check everything
		cue.Definitions(true),
		cue.Hidden(true),
		cue.Optional(true),
	)
}

func strict(v cue.Value) error {
	return v.Validate(
		// ensure final and concrete
		cue.Final(),
		cue.Concrete(true),
		// check everything
		cue.Definitions(true),
		cue.Hidden(true),
		cue.Optional(true),
	)
}
```

```
$ go run validate.go 
loose error:
#d: conflicting values int and "bar" (mismatched types int and string):
    7:5
    8:5
_h: conflicting values int and "hidden" (mismatched types int and string):
    5:5
    6:5

every error:
#d: conflicting values int and "bar" (mismatched types int and string):
    7:5
    8:5
_h: conflicting values int and "hidden" (mismatched types int and string):
    5:5
    6:5

strict error:
#d: conflicting values int and "bar" (mismatched types int and string):
    7:5
    8:5
_h: conflicting values int and "hidden" (mismatched types int and string):
    5:5
    6:5


value:
i: int
s: string
t: {
        [string]: string
}
_h: _|_ // conflicting values int and "hidden" (mismatched types int and string)
#d: _|_ // conflicting values int and "bar" (mismatched types int and string)
```

---

I understand how this is currently working. I'm starting a discussion with an example to discuss what should be done.

1. Documentation updated to reflect which options are supported.
2. Perhaps different option types are created so there is no ambiguity.
3. Errors or warnings provided when unsupported options are provided.
4. Maybe these options could be implemented, would have to think about what the semantics means.

A use case could be to separate out different classes of errors for better feedback to end users of my own applications.
1. did the user forget to set a required value in the "public" API of my definition
2. did I forget to cover all cases and the user should report an error back to the project
3. is the value usable for the upcoming computations

**Comment by antong** [comment]
I don't think the options for Validate can be used to ignore errors in hidden fields or definitions, so can't really be used like that to do "strict" or "loose" validation as in your example. Errors are errors. I guess only the Concrete option is relevant for Validate, and it is to set whether an incomplete value is an error or not.

```
package main

import (
	"fmt"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
)

func main() {
	ctx := cuecontext.New()
	t1 := ctx.CompileString(`
		a: int
		a: >=3
	`)
	for _, concrete := range []bool{true, false} {
		fmt.Printf("concrete=%v: %v\n", concrete, t1.Validate(cue.Concrete(concrete)))
	}
	// Output:
	// concrete=true: a: incomplete value >=3 & int
	// concrete=false: <nil>
}
```

**Reply by verdverm** [reply]
(moved to original description below the line)

**Reply by antong** [reply]
I agree it would be very useful to have the function documentation say which options apply, especially in cases like this where only some have any effect.

I might be totally misunderstanding things, but can Validate really work like you said you expected and not return an error based on the options to Validate()? In the example code, I'd guess the value v is already an error from the start:
```
v := ctx.CompileString(val) // val has contradictions, v is already bottom and v.Err() != nil.
```
So if this is how it is, then perhaps it is not reasonable for Validate to forgive the pre-existing errors.

**Reply by verdverm** [reply]
I could update the example CUE being processed, but there are other types of errors than just contradictions. This is more about partial evaluation / validation and the possibilities. Providing a second value to act as a "mask" is another possibility.

**Reply by myitcv** [reply]
@verdverm please can you provide some more details on the use case driving this? From memory (and some experience) the options passed to `Validate()` are due a revamp as part of revisiting the `cuelang.org/go/cue` API, but any such revamp will, as ever, be strongly driven by use cases that help shape the "right" API.
